workflows/flows/import_additional_contained_genomes_flow.py

```diff
@@ -0,0 +1,293 @@
+    rows_by_run_id: Dict[
+        int, List[Tuple[str, str, Optional[float], Optional[float]]]
+    ] = {}
+    missing_run_or_genome = 0
+    for run_acc, genome_acc, containment, cani in normalised:
+        run = runs_map.get(run_acc)
+        genome = genomes_map.get(genome_acc)
+        if run is None or genome is None:
+            missing_run_or_genome += 1
+            continue
+        rows_by_run_id.setdefault(run.id, []).append(
+            (run_acc, genome_acc, containment, cani)
+        )
+
+    if not rows_by_run_id:
+        return 0, skipped, missing_run_or_genome
+
+    run_ids = list(rows_by_run_id.keys())
+
+    # Collect all run accessions present in this chunk (for ena_accessions overlap lookup)
+    run_acc_to_ids: Dict[str, List[int]] = {}
+    for run_acc, run in runs_map.items():
+        if run.id in rows_by_run_id:
+            for acc in getattr(run, "ena_accessions", []) or []:
+                if not acc:
+                    continue
+                run_acc_to_ids.setdefault(acc, []).append(run.id)
+
+    # Find assemblies linked to any of these runs either by FK or overlapping ena_accessions
+    assemblies_qs = (
+        Assembly.objects.filter(
+            Q(run_id__in=run_ids)
+            | Q(ena_accessions__overlap=list(run_acc_to_ids.keys()))
+        )
+        .select_related(None)
+        .only("id", "run_id", "ena_accessions")
+    )
+
+    # Map run_id -> list of Assembly ids
+    assemblies_by_run_id: Dict[int, List[int]] = {rid: [] for rid in run_ids}
+    for a in assemblies_qs:
+        candidate_run_ids = set()
+        if a.run_id in assemblies_by_run_id:
+            candidate_run_ids.add(a.run_id)
+        for acc in getattr(a, "ena_accessions", []) or []:
+            for rid in run_acc_to_ids.get(acc, []):
+                candidate_run_ids.add(rid)
+        for rid in candidate_run_ids:
+            assemblies_by_run_id.setdefault(rid, []).append(a.id)
+
+    # Prepare rows to insert
+    to_create: List[AdditionalContainedGenomes] = []
+    for rid, row_list in rows_by_run_id.items():
+        assembly_ids = assemblies_by_run_id.get(rid, [])
+        if not assembly_ids:
+            continue
+        for _run_acc, genome_acc, containment, cani in row_list:
+            genome = genomes_map.get(genome_acc)
+            if genome is None:
+                continue  # already counted above, but double safety
+            for aid in assembly_ids:
+                to_create.append(
+                    AdditionalContainedGenomes(
+                        run_id=rid,
+                        genome_id=genome.genome_id,
+                        assembly_id=aid,
+                        containment=containment,
+                        cani=cani,
+                    )
+                )
+
+    created = 0
+    # Insert in DB in batches with ignore_conflicts for unique constraint
+    for batch in _chunked_iterable(to_create, batch_size):
```

Comment:
Why do you have a batching step? `bulk_create` already has a batch_size parameter?

Reply:
The file I'm trying to import has 57 million rows. 
I used both batching and bulk_create for each batch as I think we might need both for this scale. 
Do let me know your thoughts though.

Reply:
Yes, I will do that.